File: store/cxxsp/Default.cpp
#include <RequestDispatcher.h>
#include <boost/lexical_cast.hpp>
#include <boost/shared_ptr.hpp>
#include "Product.h"
#include "Category.h"
#include "Article.h"
#include "mysql.h"
#include <HttpServlet.h>

using namespace std;
using boost::lexical_cast;
using boost::shared_ptr;
using namespace servlet;

class Default : public HttpServlet {
    typedef vector<Category> categories_t;
    typedef vector<Article> articles_t;
    typedef vector<Product> products_t;

    categories_t fetch_categories();
    shared_ptr<MYSQL_RES> articles_query_product(const string& category);
    shared_ptr<MYSQL_RES> articles_query(const string& category);
    articles_t fetch_articles(const string& category);
    articles_t fetch_articles_product(const string& category);
    products_t fetch_products();
public:
    void doGet(HttpServletRequest& request, HttpServletResponse &response);
};

//-----------------------------------------------------------------------------------


Default::categories_t Default::fetch_categories() {
    const string sql = "SELECT * FROM store_categories";
    shared_ptr<MYSQL_RES> result = mysql.query(sql);

    categories_t categories;
    MYSQL_ROW row;
    while( (row = mysql_fetch_row(result.get())) ) {
        Category category;
        category.id=lexical_cast<long>(mysql.get(row,0));
        category.name=mysql.get(row,1);
        categories.push_back(category);
    }
    return categories;
}

shared_ptr<MYSQL_RES> Default::articles_query_product(const string& category) {
    const string sql = "SELECT art.article_title, art.article_desc, prod.product_id, prod.image_url FROM "
        " store_editorial_products as art, store_products as prod WHERE "
        " art.editorial_cat_id="+category+" and art.product_id=prod.product_id";
    return mysql.query(sql);
}
shared_ptr<MYSQL_RES> Default::articles_query(const string& category) {
    const string sql = "SELECT art.article_title, art.article_desc FROM "
        " store_editorial_products as art WHERE "
        " art.editorial_cat_id="+category;
    return mysql.query(sql);
}


Default::articles_t Default::fetch_articles(const string& category) {
    shared_ptr<MYSQL_RES> result = articles_query(category);
    articles_t articles;
    MYSQL_ROW row;
    while( (row = mysql_fetch_row(result.get())) ) {
        Article article;
        article.title = mysql.get(row,0);
        article.desc = mysql.get(row,1);
        articles.push_back(article);
    }
    return articles;
}
Default::articles_t Default::fetch_articles_product(const string& category) {
    shared_ptr<MYSQL_RES> result = articles_query_product(category);
    articles_t articles;
    MYSQL_ROW row;
    while( (row = mysql_fetch_row(result.get())) ) {
        Article article;
        article.title = mysql.get(row,0);
        article.desc = mysql.get(row,1);
        article.link = "ProductDetails.sxx?product_id=" + mysql.get(row,2);
        article.img = mysql.get(row,3);

        articles.push_back(article);
    }
    return articles;
}
Default::products_t Default::fetch_products() {
    shared_ptr<MYSQL_RES> result = mysql.query("SELECT * FROM store_products WHERE is_recommended");
    products_t products;
    MYSQL_ROW row;
    while( (row = mysql_fetch_row(result.get())) ) {
        Product product;
        product.product_id = lexical_cast<long>(mysql.get(row,0));
        product.category_id = lexical_cast<long>(mysql.get(row,1));
        product.name = mysql.get(row,2);
        product.price = lexical_cast<float>(mysql.get(row,3));
        product.img = mysql.get(row,4);
        product.desc = mysql.get(row,5);
        product.recomended = lexical_cast<bool>(mysql.get(row,6));
        products.push_back(product);
    }
    return products;
}



void Default::doGet(HttpServletRequest& request, HttpServletResponse &response) {
    try{
        request.setAttribute("categories", fetch_categories());
        request.setAttribute("weekly", fetch_articles("4"));
        request.setAttribute("featured", fetch_articles_product("3"));
        request.setAttribute("wereReading", fetch_articles_product("1"));
        request.setAttribute("notable", fetch_articles_product("2"));
        request.setAttribute("recomended", fetch_products());
    }catch(const exception& ex){
        request.setAttribute<string>("error",ex.what());
    }
    request.getRequestDispatcher("DefaultView.csp")->forward(request,response);
}

EXPORT_SERVLET(Default);

File: store/jsp/src/store/Default.java
package store;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;

public class Default extends HttpServlet {
    private Connection connection;

    ArrayList fetch_categories() throws SQLException {
        final String sql = "SELECT * FROM store_categories";

        ArrayList categories = new ArrayList();
        Statement statement = null;
        ResultSet result = null;
        try {
            statement = connection.createStatement();
            result = statement.executeQuery(sql);
            while( result.next() ) {
                Category category = new Category();
                category.id = Long.parseLong(result.getString(1));
                category.name=result.getString(2);
                categories.add(category);
            }
        }finally{
            Mysql.free(result,statement);
        }
        return categories;
    }        
    ResultSet articles_query_product(final String category) throws SQLException {
        final String sql = "SELECT art.article_title, art.article_desc, prod.product_id, prod.image_url FROM " +
            " store_editorial_products as art, store_products as prod WHERE " +
            " art.editorial_cat_id="+category+" and art.product_id=prod.product_id";
        return Mysql.query(connection, sql);
    }
    ResultSet articles_query(final String category) throws SQLException {
        final String sql = "SELECT art.article_title, art.article_desc FROM " +
            " store_editorial_products as art WHERE " +
            " art.editorial_cat_id="+category;
        return Mysql.query(connection, sql);
    }
    ArrayList fetch_articles(final String category) throws SQLException {
        ResultSet result = articles_query(category);
        ArrayList articles = new ArrayList();
        while( result.next() ) {
            Article article = new Article();
            article.title = result.getString(1);
            article.desc = result.getString(2);
            articles.add(article);
        }
        return articles;
    }
    ArrayList fetch_articles_product(final String category) throws SQLException {
        ResultSet result = articles_query_product(category);
        ArrayList articles = new ArrayList();
        while( result.next() ) {
            Article article = new Article();
            article.title = result.getString(1);
            article.desc = result.getString(2);
            article.link = "ProductDetails.sxx?product_id=" + result.getString(3);
            article.img = result.getString(4);
            
            articles.add(article);
        }
        return articles;
    }
    ArrayList fetch_products(Connection connection) throws SQLException {
        ResultSet result = Mysql.query(connection, "SELECT * FROM store_products WHERE is_recommended");
        ArrayList products = new ArrayList();
        while( result.next() ) {
            Product product = new Product();
            product.product_id = Long.parseLong(result.getString(1));
            product.category_id = Long.parseLong(result.getString(2));
            product.name = result.getString(3);
            product.price = Float.parseFloat(result.getString(4));
            product.img = result.getString(5);
            product.desc = result.getString(6);
            product.recomended = Boolean.getBoolean(result.getString(7));
            products.add(product);
        }
        return products;
    }
/*
    public void init() throws ServletException {
        connection = Mysql.create();
    }

    public void destroy() {
        Mysql.destroy(connection);
    }
*/

    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
        connection = Mysql.get();
        try{
            request.setAttribute("categories", fetch_categories());
            request.setAttribute("weekly", fetch_articles("4"));
            request.setAttribute("featured", fetch_articles_product("3"));
            request.setAttribute("wereReading", fetch_articles_product("1"));
            request.setAttribute("notable", fetch_articles_product("2"));
            request.setAttribute("recomended", fetch_products(connection));
        }catch(Throwable ex){
            ex.printStackTrace();
            request.setAttribute("error", ex.getMessage());
        } 
        request.getRequestDispatcher("DefaultView.jsp").forward(request,response);
    }
};
